apps/deduplication/tasks/indexing.py

```diff
@@ -0,0 +1,148 @@
+
+logger = get_task_logger(__name__)
+
+
+def find_and_set_duplicate_leads(index: MinHashLSH, lead: Lead, minhash: LeanMinHash):
+    duplicate_lead_ids = index.query(minhash)
+    duplicate_leads = Lead.objects.filter(pk__in=duplicate_lead_ids)
+    lead.duplicate_leads.set(duplicate_leads)
+
+
+def process_and_index_lead(lead: Lead, index: MinHashLSH):
+    text = lead.leadpreview.text_extract if hasattr(lead, 'leadpreview') else lead.text
+    if not text:
+        return index
+    minhash = get_minhash(text)
+    find_and_set_duplicate_leads(index, lead, minhash)
+
+    insert_to_index(index, lead.id, minhash)
+    lead.is_indexed = True
+    lead.indexed_at = timezone.now()
+    lead.save(update_fields=['is_indexed', 'indexed_at'])
+    return index
+
+
+def process_and_index_leads(
+    project: Project,
+    index_obj: LSHIndex,
+):
+    # Fetch leads which have been extracted and which have not been indexed
+    leads = Lead.objects.filter(
+        project=project,
+        is_indexed=False,
+    )
+    index: MinHashLSH = index_obj.index
+    try:
+        batches = batched(leads, batch_size=200)
+        for batch in batches:
+            with transaction.atomic():
+                for lead in batch:
+                    process_and_index_lead(lead, index)
+                # Update the index
+                index_obj.index = index
+                index_obj.save()
+    except Exception:
+        logger.warning(
+            f"Error creating index for project {project.title}({project.id})",
+            exc_info=True
+        )
+        import traceback
+
+        index_obj.has_errored = True
+        index_obj.error = traceback.format_exc()
+        index_obj.save(update_fields=['has_errored', 'error'])
+    else:
+        index_obj.status = LSHIndex.IndexStatus.CREATED
+        index_obj.save(update_fields=['status'])
+
+
+def create_project_index(project: Project):
+    """
+    Index all the unindexed leads in the project
+    """
+    index_obj = get_index_object_for_project(project)
+
+    if index_obj.has_errored:
+        logger.warning(f"Errored index object, LSHIndex id {index_obj.id}. Aborting.")
+        return
+
+    process_and_index_leads(project, index_obj)
+
+
+@shared_task
+def create_indices():
+    for project in Project.objects.all():
+        create_project_index(project)
+
+
+def get_index_object_for_project(project: Project) -> LSHIndex:
+    index_obj, created = LSHIndex.objects.get_or_create(
+        project=project,
+        defaults={
+            "name": project.title,
+            "pickle_version": pickle.format_version,
+        },
+    )
+
+    if created:
+        # MinHashLSH object will not be initialized during the creation, create it
+        index = MinHashLSH(
+            threshold=LSHIndex.THRESHOLD,
+            num_perm=LSHIndex.NUM_PERM,
+        )
+        index_obj.index = index
+        index_obj.save()
+    return index_obj
+
+
+@shared_task
+@warn_on_exception(logger)
```

Comment:
This way we might miss errors. Shouldn't we handle them instead?
Also, the sentry will skip `logger.warning`.

Reply:
Changed so that `logger.error` is called.